rfcs/0169-feature-parameter-names.md

```diff
@@ -0,0 +1,595 @@
+compared to already existing systems.
+
+1. Writing generic code is complicated. Even if list of packages all have
+   optional dependency on package `foo`, duck-typing like
+
+```
+map (p: p.override { with_foo = false; }) packages
+```
+
+won't work because of lack of uniformity, leading to ugly code downstream.
+
+2. If user wants to globally disable support for optional feature or build
+   dependency as much as possible, he needs to set multiple feature parameters,
+   and he still can miss some. E.g, it is not obvious whether the following
+   list is enough to cover everything X11-related:
+
+```
+withX
+withX11
+x11Support
+enableX
+```
+
+3. There is no clear separation between feature parameters and other arguments
+   to the package. With naming mandated by this RFC, distinction is clear to
+   cater to automatic queries.
+
+4. Side effect of the proposdd migration plan is compiling of almost
+   exahustive list of known feature parameters.
+
+
+# Detailed design
+[design]: #detailed-design
+
+When upstream build system has options to enable or disable a optional features
+or configure some compile-time parameter, like path to default mailbox or
+initial size of hash map, derivation function (for short, package) MAY expose
+compile-time options via feature parameters that match
+`^(with|conf|enable)_[a-z_0-9]$` regular expression. For short, we will refer
+to them as `with`-parameter, `conf`-parameter and `enable`-parameter correspondingly,
+collectively as `feature parameters`.
+
+Note that this RFC does not establish any requirements whether package should or
+should not have any feature parameters, this decision is left at descretion of
+the package maintainer(s). This RFC only concerns itself with the naming of the
+feature flags. Further text will use wording "corresponding feature parameter"
+with implicit caveat "provided that package maintainer decided to export
+underlying build system configuration option".
+
+Feature parameters MUST only be used according to the rules outlined in this
+RFC. They MUST NOT be used for any other purpose.
+
+Boolean feature parameter MUST be either `with`-parameter or `enable`-parameter.
+Non-boolean (e.g. string or integer) feature parameter MUST be an `conf`-parameter.
+
+Feature parameter that incurs additional runtime dependency MUST be a `with`-parameter.
+
+<details><summary>Example</summary>
+
+```
+{ lib, stdenv, mpfr, with_mpfr ? true }:
+
+stdenv.mkDerivation {
+  # ... snip ...
+  configureFlags = lib.optionals with_mpfr ["--with-mpfr"];
+  buildInputs = lib.optionals with_mpfr [ mpfr ];
+  # ... snip ...
+}
+```
+
+</details>
+
+As special provision, if optional vendored dependency is exposed by feature
+parameter, it MUST be `with`-parameter.
+
+If feature parameter does not incur extra runtime dependencies, corresponding
+feature parameter MUST be `enable`-parameter and its name SHOULD correspond to
+the upstream naming, unless is causes major inconsistency with other packages.
```

Comment:
Some projects use build systems that don’t support feature flags but do support customization via environment variables or preprocessor definitions (e.g., makefiles or xcodebuild). Am I correct assuming there be some leeway when translating their flags to appropriate feature flags in nixpkgs?

For example, MoltenVK has an [option](https://github.com/KhronosGroup/MoltenVK?tab=readme-ov-file#metal_private_api) to determine whether to use private APIs. That option is named `MVK_USE_METAL_PRIVATE_API`. A direct translation would be `enable_mvk_use_metal_private_api`, which is verbose and funky. The current package uses [`enablePrivateAPIUsage`](https://github.com/NixOS/nixpkgs/blob/b06025f1533a1e07b6db3e75151caa155d1c7eb3/pkgs/os-specific/darwin/moltenvk/default.nix#L23), which I assume would translate to something like `enable_private_api_usage` or `enable_metal_private_api_usage`.

Reply:
Yes, consistency across nixpkgs is more important than matching upstream naming exactly (I think I have this line somewhere).

I think we can call this option `enable_private_api`, and use this name for other, similar situations.
I'll add your example into guidelines.